### PR TITLE
Document stdlib; add a few missing functions; consistent naming

### PR DESCRIPTION
**Pull request by ErinvanderVeen**
Related to  #321

### CODE REVIEW

File: stdlib/strings.ncl [hunk 1]
```diff
@@ -51,43 +121,305 @@
...
+      "#m
+    = fun n => %toStr% n,
 
-    fromNum | Num -> Str = fun n => %toStr% n,
     // fromEnum | < | Dyn> -> Str = fun tag => %toStr% tag,
-    fromEnum | Dyn -> Str = fun tag => %toStr% tag,
-    fromBool | Bool -> Str = fun b => %toStr% b,
+    fromEnum | Dyn -> Str
+    | doc m#"
+      Converts any basic value to its string representation.
```
Comment (yannham) [inline stdlib/strings.ncl:363]: It's technically true, but it's an implementation detail: this function is for enums only. We should probably make it an error to use it with anything else than enum by using a more specific contract for the input. For the time being, let's just lie in the documentation :stuck_out_tongue:

File: stdlib/strings.ncl [hunk 2]
```diff
@@ -51,43 +121,305 @@
...
+      ```nickel
+        fromBool true =>
+          "true"
+      ```
+      "#m
+    = fun b => %toStr% b,
+
+    toNum | #NumLiteral -> Num
+    | doc m#"
+      Convers a string that represents an integer to that integer.
```
Comment (yannham) [inline stdlib/strings.ncl:387]:
```suggestion
      Converts a string that represents an integer to that integer.
```

File: stdlib/strings.ncl [hunk 3]
```diff
@@ -51,43 +121,305 @@
...
+      ```nickel
+        toNum "123" =>
+          123
+      ```
+      "#m
+    = fun s => %numFromStr% s,
+
+    toBool | #BoolLiteral -> Bool
+    | doc m#"
+      Converts a string that represents a bool to that bool.
```
Comment (yannham) [inline stdlib/strings.ncl:399]:
```suggestion
      Converts a string that represents a boolean to that boolean.
```

File: stdlib/strings.ncl [hunk 4]
```diff
@@ -51,43 +121,305 @@
...
+      "#m
+    = fun s => s == "true",
 
-    toNum | #NumLiteral -> Num = fun s => %numFromStr% s,
-    toBool | #BoolLiteral -> Bool = fun s => s == "true",
     // toEnum | #Ident -> < | Dyn> = fun s => %enumFromStr% s,
-    toEnum | #Ident -> Dyn = fun s => %enumFromStr% s,
+    toEnum | #Ident -> Dyn
+    | doc m#"
+      Converts any string that represents an identifier to an enum variant.
```
Comment (yannham) [inline stdlib/strings.ncl:415]:
```suggestion
      Converts any string that represents an enum variant to that enum variant.
```

The thing is that valid enum variants and valid identifiers are the same thing. I agree it's not clear to have this contract named `Ident`. Maybe we should rename it to `EnumVariant`, `Enum` or `EnumTag` as it seems to be actually used as such.

File: stdlib/lists.ncl [hunk 5]
```diff
@@ -1,76 +1,263 @@
...
+        ```nickel
+          concat [ 1, 2, 3 ] [ 4, 5, 6 ] =>
+            [ 1, 2, 3, 4, 5, 6 ]
+        ```
+        "#m
+      = fun l1 l2 => l1 @ l2,
+
+    foldl : forall a b. (a -> b -> a) -> a -> List b -> a
+      | doc m#"
+        Fold a left-associative function over a list.
```
[the remark this answers is not in the document]
Comment (yannham) [reply stdlib/lists.ncl:98]: Also, we should mention that `foldl` is strict in the accumulator by default (it corresponds to `foldl'` in Haskell).

**Review by yannham — CHANGES_REQUESTED** [review]
First, thanks for this, documentation for the stdlib is very much overdue :slightly_smiling_face: 

I left a bunch of comments and nitpicking, because documentation is hard. As general remarks:

- We should try to avoid using FP/PL jargon like "bottom" or "right-associative fold" and explain things in simple terms as much as possible, as the target users of Nickel are not the same as, say, Haskell users.
- It feels descriptions could sometimes be made more concise by removing unnecessary wording or changing the structure while keeping all the important information. As an example with `typeOf`:
  > Results in a value representing the type of the given dynamically typed value.
  =>
  > Return a representation of the type of the given value.

  Another solution, when you need to refer to the arguments is to give them names (adapted from Haskell documentation):
  > `map f l` is the list obtained by applying `f` to each element of `l`
- The CI is failing because the code is not formatted. To solve this, please run `cargo fmt` at the root of the repo before committing changes, or in a separate formatting commit once you're done.

File: stdlib/lists.ncl (hunk 5 above)
[the remark this answers is not in the document]
Comment (ErinvanderVeen) [reply stdlib/lists.ncl]: I solved this by removing any notion of associativity, adding the formula as you suggested, and adding an extra rewrite step to the example.

**Review by yannham — APPROVED** [review]

**Pull request merged**